Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h

```diff
@@ -50,128 +53,109 @@ enum class RestrictedOpenerType : uint8_t;
+template<typename DerivedType, typename BackingDataType>
+class ListDataController : public ListDataControllerBase {
 public:
-    static Ref<StorageAccessPromptQuirkObserver> create(Function<void()>&& callback)
+    static DerivedType& shared()
     {
-        return adoptRef(*new StorageAccessPromptQuirkObserver(WTFMove(callback)));
+        static MainThreadNeverDestroyed<DerivedType> sharedInstance;
+        return sharedInstance.get();
     }
 
-    ~StorageAccessPromptQuirkObserver() = default;
-
-    void invokeCallback() { m_callback(); }
-
-private:
-    explicit StorageAccessPromptQuirkObserver(Function<void()>&& callback)
-        : m_callback { WTFMove(callback) }
-    {
-    }
-
-    Function<void()> m_callback;
-};
-
-class StorageAccessUserAgentStringQuirkObserver : public RefCounted<StorageAccessUserAgentStringQuirkObserver>, public CanMakeWeakPtr<StorageAccessUserAgentStringQuirkObserver> {
-public:
-    static Ref<StorageAccessUserAgentStringQuirkObserver> create(Function<void()>&& callback)
+    void setCachedListDataForTesting(BackingDataType&& data)
     {
-        return adoptRef(*new StorageAccessUserAgentStringQuirkObserver(WTFMove(callback)));
+        m_wasInitialized = true;
+        setCachedListData(WTFMove(data));
+        m_observers.forEach([](auto& observer) {
+            observer.invokeCallback();
+        });
     }
 
-    ~StorageAccessUserAgentStringQuirkObserver() = default;
+    const BackingDataType& cachedListData() const { return m_cachedListData; }
 
-    void invokeCallback() { m_callback(); }
+protected:
+    friend class NeverDestroyed<DerivedType, MainThreadAccessTraits>;
 
-private:
-    explicit StorageAccessUserAgentStringQuirkObserver(Function<void()>&& callback)
-        : m_callback { WTFMove(callback) }
+    void setCachedListData(BackingDataType&& data)
     {
+        m_cachedListData = WTFMove(data);
+        didUpdateCachedListData();
     }
 
-    Function<void()> m_callback;
+    virtual void didUpdateCachedListData() { }
+
+    BackingDataType m_cachedListData;
 };
 
-class StorageAccessPromptQuirkController {
+class LinkDecorationFilteringController : public ListDataController<LinkDecorationFilteringController, Vector<WebCore::LinkDecorationFilteringData>> {
 public:
-    static StorageAccessPromptQuirkController& shared();
-
-    const Vector<WebCore::OrganizationStorageAccessPromptQuirk>& cachedQuirks() const { return m_cachedQuirks; }
-    void updateQuirks(CompletionHandler<void()>&&);
-    void setCachedQuirksForTesting(Vector<WebCore::OrganizationStorageAccessPromptQuirk>&&);
-    void initialize();
-
-    Ref<StorageAccessPromptQuirkObserver> observeUpdates(Function<void()>&&);
+    void updateList(CompletionHandler<void()>&&) final;
+    const Vector<WebCore::LinkDecorationFilteringData>& cachedListData() const { return m_cachedListData; }
```

Comment:
You shouldn't need this definition given `const BackingDataType& cachedListData() const { return m_cachedListData; }`

Reply:
Whoops, meant to remove this — good catch!